8632.6 combine ledger fix tests add documentation and tidy up

andreievg opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Fixes #8632 (partially)
Fixes https://github.com/msupply-foundation/open-msupply/issues/8431
Fixes https://github.com/msupply-foundation/open-msupply/issues/8675

# 👩🏻‍💻 What does this PR do?

* Sorry I snuck in the fix for the label print per line disabled in verified (sorry did this while waiting for compilation, might as well have it ? Or should I do another PR?)
* Adds documentation
* Combines all ledger mock data in one test (make sure they behave together)
* Improves loggin by adding progress
* Deals with -0.0
* Increments omSupply version compatibility

# Testing

One test case is here: https://docs.google.com/document/d/1qIBm7Iaf1RZgs_CkrJDm-4j0oN-aXwq3iVncow1xVtM/edit?tab=t.0

Try upgrading a site that has historic ledger discrepancies, wait for ledger fix to finish (can check in logs), see ledger discrepancy fixed.

Also system log will have the log after remote site fixed.

In terms of seep, the test that I've done on a reasonably large data set (200k records, not sure how many invoice_lines), with 1k ledger discrepancies on the samsung tablet, it took about 4 minutes (I would suspect with data sets getting large it's more of exponential slow down)

Chris-Petty commented on server/service/src/rnr_form/generate_rnr_form_lines.rs:
```diff
@@ -133,7 +134,7 @@ pub fn generate_rnr_form_lines(
                 previous_monthly_consumption_values,
                 average_monthly_consumption,
-                initial_balance,
+                initial_balance: pos_zero(initial_balance),
```
Roxy did notice this in earlier work but I couldn't reproduce. I was hoping if we did reproduce we'd fix it at the DB level otherwise anything depending on the source of the issue will have to bodge it.

I suspect `get_opening_balance` which gets the initial balance could only be returning `-0.0` if the previous final balance is negative or `StockOnHandRepository` is returning `-0.0`.

(deduced from in rust `0.0 - 0.0` is `0.0`, and `-0.0 - 0.0` is `-0.0`)

Chris-Petty commented on server/service/src/ledger_fix/README.md:
```diff
@@ -0,0 +1,103 @@
+
+Due to migration from existing mSupply sites and omSupply bug we've had ledger discrepancies, which are:
+* Where total movements of stock (invoice_lines) does not add up to current stock balance (stock_line)
+* Where as any point in history, sum of invoice_lines as negative.
```
```suggestion
* Where at any point in history, sum of invoice_lines as negative.
```

Chris-Petty commented on server/service/src/ledger_fix/README.md:
```diff
@@ -0,0 +1,103 @@
+
+* Once a day
+* After initialisation
+* After upgrade to this version
```
Just make it clearer when this was introduced in the far future

```suggestion
* After upgrading from a version < v2.9.2 to a version >= v2.9.2
```

Chris-Petty commented on server/service/src/ledger_fix/README.md:
```diff
@@ -0,0 +1,103 @@
+
+### Fix Ledger discrepancies
+
+Each stock line will go through a series of fixes based on know 'states' of ledger. After each fix a ledger discrepancy check is performed, if ledger is not yet fixed the next fix is attempted.
```
```suggestion
Each stock line will go through a series of fixes based on known invalid of ledger states. After each fix a ledger discrepancy check is performed, if ledger is not yet fixed the next fix is attempted.

If a ledger cannot be corrected by the implemented automatic fixes, then our team should address them with after automated notification of their occurance.
```

Chris-Petty commented on server/service/src/ledger_fix/mod.rs:
```diff
@@ -0,0 +1,87 @@
+    reserved_not_picked: f64,
+}
+
+pub(crate) fn ledger_balance_summary(
```
nit - could be `LedgerBalanceSummary::new(...)`

andreievg replied:
With "new" we loose sense of order of f64 fields (since they would all just be params separated by ',' easy to make mistake), i typically think new is not very useful in most cases

Chris-Petty replied:
I don't follow isn't it just:
```rs
let balance_summary = ledger_balance_summary(connection, &ledger_lines, stock_line_id)?;
// vs 
let balance_summary = LedgerBalanceSummary::new(connection, &ledger_lines, stock_line_id)?;
```
Note this truly doesn't matter just trying to figure out why sometimes we favour one over another!

Chris-Petty commented on server/repository/src/db_diesel/stock_line_ledger_discrepancy.rs:
```diff
@@ -0,0 +1,82 @@
+    let mut query = stock_line_ledger_discrepancy::table.into_boxed();
+
+    let Some(StockLineLedgerDiscrepancyFilter {
+        stock_line,
```
IDK if just me but I find the naming of this confusing. Without context I assume it's a `StockLine` struct at which point why is stock_line_id here separately. But it actually is a StockLineFilter... but we don't conventionally have struct field names in our filter structs that have the keyword "Filter" i.e. `StockLineLedgerDiscrepancyFilter::stock_line_filter: Option<StockLineFilter>`. Though I do think that'd be clearer... idk.

TBF I think it's clearer with IDE inlay hints, but you don't get that in github ;-). Maybe inlay hints make us not think through naming as much haha.

andreievg replied:
We have had this pattern before, I'll make carry over to remove stock_line filter and just add store_id to StockLineLedgerDiscrepancy which is all we need atm tbh, thanks

Chris-Petty reviewed (commented):
23/32. Getting to the meat of the fix code next

andreievg commented on server/repository/src/db_diesel/stock_line.rs:
```diff
@@ -175,6 +175,77 @@ impl<'a> StockLineRepository<'a> {
         Ok(result.into_iter().map(to_domain).collect())
     }
+
+    pub fn create_filtered_query(filter: Option<StockLineFilter>) -> BoxedStockLineQuery {
```
Yep, sorry I though I comment on that PR should have made a comment that it was just moved

andreievg replied:
Actually this change shouldn't be there anymore, it was committed in earlier PR and I think you reviewed it (i think i added a comment that it was moved there), sorry with the squash commit this is more difficult

Chris-Petty commented on server/service/src/ledger_fix/fixes/adjust_all_to_match_available.rs:
```diff
@@ -0,0 +1,172 @@
+            make_movements(nothing_matches, vec![(2, 10), (4, -10), (5, 30), (6, -7)]);
+
+        // Add reserved not picked
+        nothing_matches.invoices[3] = inline_edit(&nothing_matches.invoices[3], |mut u| {
+            u.status = InvoiceStatus::Allocated;
+            u.picked_datetime = None;
+            u.shipped_datetime = None;
+            u.received_datetime = None;
+            u.verified_datetime = None;
+            u
+        });
```
Never mind comments on this. Will play around to see if I can get an agent to PR refactoring inline_edit outta the codebase another day.

andreievg replied:
Haha, thanks, happy to not use inline_edit, actually you so right in this case it doesn't even make sense

andreievg commented on server/service/src/ledger_fix/fixes/adjust_all_to_match_available.rs:
```diff
@@ -0,0 +1,172 @@
+        ..
+    } = balance_summary;
+
+    let should_adjust = available + reserved_not_picked != total
```
Sure, the reason I didn't go with allocated was because available stock remains allocated after allocated statues. So something that is allocated but not taken out of total ?

andreievg commented on server/service/src/ledger_fix/fixes/mod.rs:
```diff
@@ -0,0 +1,211 @@
+    new_end_balance: f64,
+    stock_line_id: &str,
+) -> Result<(), LedgerFixError> {
+    let adjustment = (current_end_balance - new_end_balance) * -1.0;
```
thanks yeah! ad6090456eb1b4bd8402d2b0eeaa35aeace899af, compiles

andreievg commented on server/service/src/ledger_fix/fixes/mod.rs:
```diff
@@ -0,0 +1,211 @@
+        return Ok(());
+    }
+
+    // if adjustment is negative we need to add negative adjustment as far back in history as possible (without causing negative balance)
```
yep, but at the same time wanted to reduce impact to user (or at least try to), I made an assumption that the closer to current date the adjustment is, the more likely the impact.

andreievg commented on server/service/src/test_helpers.rs:
```diff
@@ -115,3 +118,70 @@ pub mod email_test {
+                        InvoiceStatus::Verified
+                    } else {
+                        use repository::InvoiceStatus;
+
```
thanks: 4e4af9c8f0081c83df022b0c42592307d372a428

andreievg commented on server/service/src/ledger_fix/find_ledger_discrepancies.rs:
```diff
@@ -0,0 +1,172 @@
+            ],
+        );
+
+        allocated_not_picked_movements.invoices[2] =
```
You can ask you to write a script that does it : ), i always wondered if that's the next evolution

andreievg commented:
Thank you so much @Chris-Petty ! 

> Thanks for review so far @Chris-Petty ! I think i dealt with all comments, fixed conflicts (which reduced diff quite a bit) and fixed tests

Tests pass now : )
